picard/ui/metadatabox/__init__.py

```diff
@@ -265,41 +295,134 @@ def keyPressEvent(self, event):
+                        value = value.split(MULTI_VALUED_JOINER)
+                        yield from self._set_tag_values_delayed_updates(tag, value)
+                    else:
+                        log.error("Tag '%s' without new or old value found in clipboard, ignoring.", tag)
+
+        data = _decode_json(mimedata)
+        return _apply_tag_dict(data) if data else []
+
+    def _paste_from_text(self, mimedata):
         item = self.currentItem()
-        if item:
-            column = item.column()
-            tag = self.tag_diff.tag_names[item.row()]
-            value = None
-            if column == self.COLUMN_ORIG:
-                value = self.tag_diff.old[tag]
-            elif column == self.COLUMN_NEW:
-                value = self.tag_diff.new[tag]
-
-            if tag == '~length':
-                try:
-                    value = [format_time(value or 0), ]
-                except (TypeError, ValueError) as why:
-                    log.warning(why)
-                    value = ['']
-
-            if value is not None:
-                self.tagger.clipboard().setText(MULTI_VALUED_JOINER.join(value))
-                self.clipboard = value
+        column_is_editable = (item.column() == self.COLUMN_NEW)
+        tag = self.tag_diff.tag_names[item.row()]
+        value = mimedata.text()
+        if column_is_editable and self._tag_is_editable(tag) and value:
+            log.info("Pasting %s from text clipboard to tag %s", value, tag)
+            value = value.split(MULTI_VALUED_JOINER)
+            yield from self._set_tag_values_delayed_updates(tag, value)
+
+    def _can_paste(self):
+        has_valid_mime_data = False
+        mimedata = self.tagger.clipboard().mimeData()
+        for mimetype, converters in self.mimedatas.items():
+            if mimedata.hasFormat(mimetype) and converters.decode_func:
+                has_valid_mime_data = True
+
+        return has_valid_mime_data and len(self.tracks) <= 1 and len(self.files) <= 1
 
     def _paste_value(self):
-        item = self.currentItem()
-        if item:
-            column = item.column()
-            tag = self.tag_diff.tag_names[item.row()]
-            if column == self.COLUMN_NEW and self._tag_is_editable(tag):
-                self._set_tag_values(tag, self.clipboard)
-                self.update()
+        if not self._can_paste():
+            msg = N_("No valid data in clipboard to paste")
+            self.tagger.window.set_statusbar_message(msg, echo=log.info, timeout=3000)
+            return
 
-    def _update_clipboard(self):
-        clipboard = self.tagger.clipboard().text().split(MULTI_VALUED_JOINER)
-        if clipboard:
-            self.clipboard = clipboard
+        objects_to_update = set()
+        mimedata = self.tagger.clipboard().mimeData()
+
+        for mimetype, converters in self.mimedatas.items():
+            if mimedata.hasFormat(mimetype) and converters.decode_func:
```

Comment:
We have same code above, we can convert this to a generator:

```python

def decode_funcs(self, mimedata):
    for mimetype, converters in self.mimedatas.items():
        if mimedata.hasFormat(mimetype) and converters.decode_func:
             yield converters.decode_func
```

```python
for decode_func in decode_funcs(mimedata):
    objs = decode_func(self, mimedata)
```


And above:

```python
has_valid_mime_data = any(decode_funcs(mimedata))
```

Reply:
I'm thinking of breaking this out into a TagDiffMimeRegistry class in the TagDiff file. Can add tests around it, and avoids bloating the metadata box class any more!

Reply:
Something like this? https://github.com/StevilKnevil/picard/pull/3